docs/auxil/admonition_inserter.py

```diff
@@ -0,0 +1,507 @@
+            return f":meth:`telegram.Bot.{method_name}`"
+        elif base_class == telegram.ext.ApplicationBuilder:
+            return f":meth:`telegram.ext.ApplicationBuilder.{method_name}`"
+        else:
+            raise NotImplementedError(f"Base class {base_class} not supported")
+
+    def _resolve_arg_and_add_link_to_method(
+        self,
+        arg: Any,
+        dict_of_methods_for_class: defaultdict,
+        link_to_method: str,
+        include_subclasses: bool = True,
+    ) -> None:
+        """A helper method for "Returned in" and "Use in" admonition.
+        Tries to resolve the arg to a valid class. In case of success, adds the link to
+        Bot's or ApplicationBuilder's method to the set of methods for that class in the dictionary
+        of admonitions.
+
+        If `include_subclasses_if_applicable` is set to `True`, the method will check
+        if the resolved class is included in the list of classes whose subclasses can be
+        included in the same admonition. If it is on that list, dictionary entries for its
+        subclasses are added/updated too.
+
+        **Modifies dictionary in place.**
+        """
+        for cls in self._resolve_arg(arg):
+
+            # When trying to resolve an argument from args or return annotation,
+            # the method _resolve_arg returns None if nothing could be resolved.
+            if cls is None:
+                continue
+
+            dict_of_methods_for_class[cls].add(link_to_method)
+
+            if not include_subclasses:
+                continue
+
+            for subclass in [
+                # exclude private classes
+                p
+                for p in cls.__subclasses__()
+                if not str(p).split(".")[-1].startswith("_")
+            ]:
+                dict_of_methods_for_class[subclass].add(link_to_method)
+
+    def _resolve_arg(self, arg: Any) -> Iterator[Union[type, None]]:
+        """Analyzes an argument of a method and recursively yields classes that the argument
+        or its sub-arguments (in cases like Union[...]) belong to, if they can be resolved to
+        telegram or telegram.ext classes.
+        """
+
+        origin = typing.get_origin(arg)
+
+        if (
+            origin in (collections.abc.Callable, typing.IO)
+            or arg is None
+            # no other check available (by type or origin) for these:
+            or str(type(arg)) in ("<class 'typing._SpecialForm'>", "<class 'ellipsis'>")
+        ):
+            pass
+
+        # RECURSIVE CALLS
+        # for cases like Union[Sequence....
+        elif typing.get_origin(arg) in (
+            dict,
+            tuple,
+            type,
+            Union,
+            collections.abc.Coroutine,
+            collections.abc.Sequence,
+            typing.Sequence,
+        ):
+            for sub_arg in typing.get_args(arg):
+                yield from self._resolve_arg(sub_arg)
+
+        elif isinstance(arg, list):
```

Comment:
for some reason `get_origin()` of a list returns `None`

Reply:
> for some reason `get_origin()` of a list returns `None`

hmm, well I found the same behaviour on tuples and dicts then. on py3.11:

``` python
import typing
a = typing.get_origin([1, 2, 3])  # None
b = typing.get_origin((1, 2, 3))  # None
c = typing.get_origin({1: 2})  # None
```

Reply:
I think I'll just do the opposite thing: I'll remove the types and origins that are practically unused and create a more informative exception message so that if someone creates a new argument and Sphinx fails, the user would know exactly what happened.

Reply:
okay, so the docs passing means those conditions weren't actually being hit before, and was intended as a failsafe.

> more informative exception message

nice idea